File: Programmers/[17686] 파일명 정렬/SUbbb/17686.java
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

class File {
    // 변환하지 않은 원본 파일명
    String file;
    String head;
    int number;

    public File(String file) {
        this.file = file;
        file = file.toLowerCase();

        // 숫자가 연속이거나, 숫자가 아닌 문자가 연속인 경우 찾기
        Pattern p = Pattern.compile("\\d+|\\D+");
        Matcher matcher = p.matcher(file);

        matcher.find();
        this.head = matcher.group(0);
        matcher.find();
        this.number = Integer.parseInt(matcher.group(0));
    }
}

class Solution {
    public String[] solution(String[] files) {
        ArrayList<File> fileList = new ArrayList<>();

        for (String file : files)
            fileList.add(new File(file));

        fileList.sort(((o1, o2) -> {
            // head가 동일하다면, number 기준 오름차순 정렬
            if (o1.head.compareTo(o2.head) == 0) return o1.number - o2.number;
            // head 기준 사전순 정렬
            return o1.head.compareTo(o2.head);
        }));
        
        String[] answer = new String[files.length];
        
        for (int i = 0; i < fileList.size(); i++) 
            answer[i] = fileList.get(i).file;
        
        return answer;
    }
}

File: Programmers/[17686] 파일명 정렬/JFe/17686.py
import re
def solution(files):
    answer, file_list = [], []
    for i, file in enumerate(files):
        # head 매치
        pattern = re.compile('[^0-9]+')
        m = pattern.match(file)
        head = m.group().lower()
        # number 매치
        pattern = re.compile('[0-9]+')
        m = pattern.match(file[m.end():])
        number = int(m.group())
        # file_list에 저장
        file_list.append([i, head, number, file])
    # head, number, index 순으로 정렬
    file_list = sorted(file_list, key = lambda x: (x[1], x[2], x[0]))
    # 정렬 된 순서대로 answer에 추가
    for i in range(len(file_list)):
        answer.append(file_list[i][3])
    return answer